### PR TITLE
[ROX-12211] Optimize image graphQL resolver

### PR DESCRIPTION
**Pull request by md2119**
## Description
Optimize image with scan and image without scan graphQL paths. In the past we always loaded the full image even when scan details were not needed. This PR optimizes boths paths for the new graphQL endpoints.

Note: It would be easier to review individual commits.

## Checklist
- [ ] Investigated and inspected CI test results
- [X] Unit test and regression tests added
- ~Evaluated and added CHANGELOG entry if required~
- ~Determined and documented upgrade steps~
- ~Documented user facing changes~

If any of these don't apply, please comment below.

## Testing Performed
```
BenchmarkImageResolver
    env_isolator.go:41: EnvIsolator: Setting ROX_POSTGRES_DATASTORE to true
BenchmarkImageResolver/GetImageComponentsInImageScanResolver
BenchmarkImageResolver/GetImageComponentsInImageScanResolver-8         	      19	  58440569 ns/op
BenchmarkImageResolver/GetImageComponentsWithoutImageScanResolver
BenchmarkImageResolver/GetImageComponentsWithoutImageScanResolver-8    	       3	 441044677 ns/op
BenchmarkImageResolver/GetImageComponentsDerivedFieldsWithImageScanResolver
BenchmarkImageResolver/GetImageComponentsDerivedFieldsWithImageScanResolver-8         	      12	 132206259 ns/op
BenchmarkImageResolver/GetImageComponentsDerivedWithoutImageScanResolver
BenchmarkImageResolver/GetImageComponentsDerivedWithoutImageScanResolver-8            	       2	 619884703 ns/op
BenchmarkImageResolver/GetImageOnly
BenchmarkImageResolver/GetImageOnly-8                                                 	    2704	    475802 ns/op
BenchmarkImageResolver/GetImageWithCounts
BenchmarkImageResolver/GetImageWithCounts-8                                           	      16	  71430526 ns/op
PASS
```

image

image

### CODE REVIEW

File: central/graphql/resolvers/image_components.go [hunk 1]
```diff
@@ -335,13 +339,37 @@ func (resolver *imageComponentResolver) ImageVulnerabilityCounter(ctx context.Co
...
+		return resolver.root.ImageVulnerabilities(resolver.withImageComponentScope(ctx), args)
+	}
+
+	query, err := args.AsV1QueryOrEmpty()
+	if err != nil {
+		return nil, err
+	}
+
+	pagination := query.GetPagination()
+	query.Pagination = nil
```
Comment (connorgorman) [inline central/graphql/resolvers/image_components.go:355]: why is pagination removed here? getImageCVEResolvers should respect the pagination

File: central/graphql/resolvers/image_components.go [hunk 2]
```diff
@@ -450,3 +509,35 @@ func (resolver *imageComponentResolver) UnusedVarSink(ctx context.Context, args
 func (resolver *imageComponentResolver) FixedIn(_ context.Context) string {
 	return resolver.data.GetFixedBy()
 }
+
+func getImageCVEResolvers(ctx context.Context, root *Resolver, os string, vulns []*storage.EmbeddedVulnerability, query *v1.Query) ([]ImageVulnerabilityResolver, error) {
```
Comment (connorgorman) [inline central/graphql/resolvers/image_components.go:513]: Is there a problem here if sorting is involved?

Comment (connorgorman) [reply central/graphql/resolvers/image_components.go:513]: Also I wonder if we should make this function deterministic if sorting is not a requirement and sort to make the results deterministic

File: central/graphql/resolvers/image_scan.go [hunk 3]
```diff
@@ -25,9 +32,54 @@ func init() {
...
 
 func (resolver *imageScanResolver) ImageComponents(_ context.Context, args PaginatedQuery) ([]ImageComponentResolver, error) {
-	return resolver.root.ImageComponents(resolver.ctx, args)
+	query, err := args.AsV1QueryOrEmpty()
+	if err != nil {
+		return nil, err
+	}
+
+	pagination := query.GetPagination()
+	query.Pagination = nil
```
Comment (connorgorman) [inline central/graphql/resolvers/image_scan.go:41]: same comment as above wrt to pagination

File: central/graphql/resolvers/image_scan.go [hunk 4]
```diff
@@ -25,9 +32,54 @@ func init() {
...
+		pv: pagination,
+	}.paginate(resolvers, err)
+	return ret.([]ImageComponentResolver), err
 }
 
 func (resolver *imageScanResolver) ImageComponentCount(_ context.Context, args RawQuery) (int32, error) {
 	return resolver.root.ImageComponentCount(resolver.ctx, args)
 }
+
+func getImageComponentResolvers(ctx context.Context, root *Resolver, imageScan *storage.ImageScan, query *v1.Query) ([]ImageComponentResolver, error) {
```
Comment (connorgorman) [inline central/graphql/resolvers/image_scan.go:54]: should sort for deterministic output

File: central/graphql/resolvers/image_scan_benchmark_test.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		b.SkipNow()
+	}
+
+	mockCtrl := gomock.NewController(b)
+	defer mockCtrl.Finish()
+	db, gormDB := setupPostgresConn(b)
+	defer pgtest.CloseGormDB(b, gormDB)
+	defer db.Close()
+
+	imageDataStore := createImageDatastore(b, mockCtrl, db, gormDB)
```
Comment (connorgorman) [inline central/graphql/resolvers/image_scan_benchmark_test.go:75]: any particular reason for all these stores we mock out risk? Seems a bit odd to be mixing mocked objects with real objects

**Review by connorgorman — COMMENTED** [review]
Left some comments

File: central/graphql/resolvers/image_scan_benchmark_test.go (hunk 5 above)
Comment (md2119) [reply central/graphql/resolvers/image_scan_benchmark_test.go:75]: We are not testing whether risk datastore. To enable testing risk datastore it's store needs to be moved out of internal which seems unnecessary for what this test's focus is.

File: central/graphql/resolvers/image_components.go (hunk 2 above)
Comment (md2119) [reply central/graphql/resolvers/image_components.go:505]: Moved pagination into `getImageCVEResolvers` and added default sorting on ID if not sort cue in incoming query.

**Review by connorgorman — APPROVED** [review]

File: central/graphql/resolvers/image_scan_test.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+func (s *ImageScanResolverTestSuite) TearDownTest() {
+	s.mockCtrl.Finish()
+}
+
+func (s *ImageScanResolverTestSuite) TearDownSuite() {
+	s.envIsolator.RestoreAll()
+}
+
+func (s *ImageScanResolverTestSuite) TestGetImagesWithScan() {
```
Comment (janisz) [inline central/graphql/resolvers/image_scan_test.go:96]: @md2119 This tests fails with race detected

```
{Failed  ==================
WARNING: DATA RACE
Read at 0x00c001daa1c8 by goroutine 99:
  github.com/stackrox/rox/central/graphql/resolvers.(*imageResolver).Id()
      /go/src/github.com/stackrox/stackrox/central/graphql/resolvers/generated.go:6239 +0x4c
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:626 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:339 +0xd7
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection.func2()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:211 +0xa1b
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:231 +0x374
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections.func1()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:81 +0x2f4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections·dwrap·2()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:82 +0x47

Previous write at 0x00c001daa1c8 by goroutine 100:
  github.com/stackrox/rox/central/graphql/resolvers.(*imageResolver).Scan()
      /go/src/github.com/stackrox/stackrox/central/graphql/resolvers/images.go:428 +0x1d7
  runtime.call32()
      /usr/local/go/src/runtime/asm_amd64.s:626 +0x48
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:339 +0xd7
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection.func2()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:211 +0xa1b
  github.com/graph-gophers/graphql-go/internal/exec.execFieldSelection()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:231 +0x374
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections.func1()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:81 +0x2f4
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections·dwrap·2()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:82 +0x47

Goroutine 99 (running) created at:
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:77 +0x212
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelectionSet()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:267 +0x478
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execList.func1()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:332 +0x387
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execList·dwrap·5()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:333 +0x47

Goroutine 100 (running) created at:
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelections()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:77 +0x212
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execSelectionSet()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:267 +0x478
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execList.func1()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:332 +0x387
  github.com/graph-gophers/graphql-go/internal/exec.(*Request).execList·dwrap·5()
      /go/pkg/mod/github.com/graph-gophers/graphql-go@v1.3.0/internal/exec/exec.go:333 +0x47
==================
    testing.go:1152: race detected during execution of test}
```
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stackrox_stackrox/2720/pull-ci-stackrox-stackrox-master-go-unit-tests/1560194439421366272

**Pull request merged**